Restructure compilation documentation

thomasspriggs opened the pull request:
This PR restructures the compilation documentation. We have received some feedback that the compilation instructions are difficult to follow for new users. This PR makes things more straight forward for new users to follow.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Update compiling document to indicate current platform status

Becuase some platforms are not tested in CI, a new user would be well
adviced to try a platform which is tested if they have choice.

thomasspriggs committed:
Move CMake build instructions above make build instructions

The complete repository can be built using fewer build step using CMake
than using the hand-written makefiles. Therefore people who are new to
building CBMC should be able to get started more quickly using the CMake
build than the make build. Therefore it makes sense for them to read
about that build system first.

thomasspriggs committed:
Update compiling documentation to refer to Visual Studio 2019

In order to match the version of visual studio and tool chain currently
tested in CI.

thomasspriggs committed:
Fix misspelling of "should" in compiling documentation

Because "shoud" is not the correct spelling.

thomasspriggs committed:
Document location of built binaries for the cmake build

Because this was previously missing and it might not have been obvious
for anyone unfamiliar with the build system.

thomasspriggs committed:
Link to CI script in compiling documentation

Because this helps document the way CBMC is currently built in CI. This
is usefull both for reference compared to local builds and for working
out why CI builds are failing on PRs.

thomasspriggs committed:
Move SOLARIS and FREEBSD sections downwards

To match the ordering in the outline at the top.

thomasspriggs committed:
Retitle Windows section to indicate cygwin usage

To distance this section from the comparitively well-maintained Visual
Studio 2019 with cmake build.

NlightNFotis commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
 
-The rest of this document is split up into three parts: compilation on Linux,
-MacOS, Windows.  Please read the section appropriate for your machine.
+# Building using CMAKE
```
Would it be possible we have `CMake` consistently capitalised?

thomasspriggs replied:
Done.

NlightNFotis commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+     [the github release page](https://github.com/lexxmark/winflexbison/releases/)
+     or through
+     [the chocolatey package manager.](https://community.chocolatey.org/packages/winflexbison3)
+     Installing strawberryperl is advised if you want to run tests on Windows.
```
Can `strawberryperl` be added into a codeblock to indicate that it's a package name as well? Or add instructions for it with chocolatey as an addendum?

thomasspriggs replied:
Done.

NlightNFotis commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+     [official download page](https://cmake.org/download) on those systems.
+     The dependencies (as listed in the relevant sections above) will also be
+     required, and should be installed using the suggested methods.
+2. Navigate to the *top level* folder of the project. This is different from
```
Maybe add the second sentence as a bullet point, or add the paths in parentheses (`cbmc` vs `cbmc/src/`) to make it a bit less confusing?

NlightNFotis commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+
+   Finally, to enable building universal binaries on macOS, you can pass the
+   flag `-DCMAKE_OSX_ARCHITECTURES=i386;x86_64`. If you don't supply this flag,
+   the build will just be for the architecture of your machine.
```
Would it be better phrased as _The build target is going to be the local machine's architecture_?

thomasspriggs replied:
I have rephrased this, but slightly differently.

NlightNFotis reviewed (approved):
LGTM, with some minor comments.

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+
 - Solaris 11
 - FreeBSD 11
 - Cygwin
```
Hmm, this is (already was!) a somewhat odd mix of operating systems and build environments. Perhaps say "Windows" instead of "Visual Studio" and drop "Cygwin"? There surely are tool chains for Linux that we are _not_ testing.

thomasspriggs replied:
I have dropped cygwin.

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+The above environments are currently tested as part of our continuous
+integration system. It separately tests both the CMake build system and the
+hand-written make files. The latest build steps being used in CI can be
+[referenced here](https://github.com/diffblue/cbmc/blob/develop/.github/workflows/pull-request-checks.yaml).
```
Should probably say "found here" instead of "referenced here?"

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+# Building using CMAKE
+
+Building with CMake is supported across Linux, MacOS X and Windows with Visual
+Studio 2019. There are also hand-written make files which can be used to build
+separate binaries independently. Usage of the hand-written make files is
```
What does "separate binaries independently" mean?

TGWDB replied:
I think this might be clearer as something like "can be used to build (sub) components of the cbmc package" or something like this? I agree the current version is not clear (I don't know if mi version is any better). But the spirit I think is we could build `goto-diff` or `jbmc` without building all the others?

tautschnig replied:
But ... that's no different with cmake, is it? You can easily build individual components by just specifying them as targets (e.g., I sometimes do `ninja -C build cbmc` when I don't care about building all the rest).

TGWDB replied:
I agree. I'm guessing at what the wording meant to convey.

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+Building with CMake is supported across Linux, MacOS X and Windows with Visual
+Studio 2019. There are also hand-written make files which can be used to build
+separate binaries independently. Usage of the hand-written make files is
+explained in a separate section. The CMake build can build the complete
```
Could that "separate section" be named/referenced here?

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+Studio 2019. There are also hand-written make files which can be used to build
+separate binaries independently. Usage of the hand-written make files is
+explained in a separate section. The CMake build can build the complete
+repository in fewer steps and supports better integration with various IDEs and
```
How did you find out that it was "fewer steps?"

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+     run `brew install cmake` to install CMake.
+   - On platforms where installing the Java Development Kit and Maven is
+     difficult, you can avoid needing these dependencies by not building
+     JBMC. Just pass `-DWITH_JBMC=OFF` to cmake in step (4) below.
```
The above commands didn't install maven or a JDK! Also, having this item right here is a bit strange for both the ones before and after are about installing operating-system specific dependencies.

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+   This command tells CMake to invoke the correct tool to run the build in the
+   `build` directory. You can also use the build back-end directly by invoking
+   `make`, `ninja`, or opening the generated IDE project as appropriate. The
+   complete set of built binaries can be found in `./build/bin` once the build
```
Nit picking: I'd probably use `build/bin/` instead of `./build/bin`.

tautschnig commented on COMPILING.md:
```diff
@@ -1,19 +1,125 @@
+   `build` directory. You can also use the build back-end directly by invoking
+   `make`, `ninja`, or opening the generated IDE project as appropriate. The
+   complete set of built binaries can be found in `./build/bin` once the build
+   is complete.
```
Perhaps add a comment about doing parallel builds in here? Most people will have multiple cores available these days, and the build process is a lot faster when using more than one core...

tautschnig reviewed (approved):
Thank you for improving the documentation!

TGWDB commented on COMPILING.md:
````diff
@@ -15,10 +15,106 @@ past, but are not actively tested:
+     xcode-select --install
+     ```
+     You shoud also install [Homebrew](https://brew.sh), after which you can
+     run `brew install cmake` to install CMake.
````
Also similar to tautshnig's comment below, why is MacOS different in style and arguments to the others. I think it would be better to have single (displayed) comment for each system and let that do it all. If users want to play around, this can (probably) be covered elsewhere. (Note, I'm reviewing by commit, so they may be a little haphazard depending on the commits.)

TGWDB commented on COMPILING.md:
```diff
@@ -15,10 +15,106 @@ past, but are not actively tested:
+# Building using hand-written make files.
+
+The rest of this section is split up based on the platform being built on. 
+Please read the section appropriate for your machine.
```
"machine" -> "platform"

TGWDB commented on COMPILING.md:
```diff
@@ -1,14 +1,19 @@
 
 - Linux
 - MacOS X
+- Microsoft Visual Studio
```
This should be >=2019, also below should probably indicate MSVS 2015+-?

thomasspriggs committed:
Quote `strawberryperl` in compiling documentation

In order to indicate that this is the name of the package.

thomasspriggs committed:
Consistently capitalise CMake when naming the build system

Because "CMake" should be capitalised that way; not "cmake" or "CMAKE".

thomasspriggs committed:
Quote `cmake` when referring to the command

To distinguish it from the name of the build system as the command
should not be capitalised.

thomasspriggs committed:
Document where to get Strawberry Perl in `COMPILING.md`

To help users who decide to install Strawberry Perl to do so.

thomasspriggs committed:
Rephrase explaination of non-universal macOS binaries

In order to more clearly explain the issue.

thomasspriggs committed:
Remove references to cygwin in favour of Windows instead

Because cygwin is not tested in CI and because this means that we can
list operating systems rather than a mix of operating systems and tool
chains.
